hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/async/RouterAsyncRpcClient.java

```diff
@@ -627,4 +697,46 @@ public <T> T invokeSingle(
     invokeSequential(locations, remoteMethod);
     return asyncReturn(clazz);
   }
+
+  protected void acquirePermit(final String nsId, final UserGroupInformation ugi,
+      final Method m, RouterRpcFairnessPolicyController controller)
+      throws IOException {
+    if (controller != null) {
+      if (!controller.acquirePermit(nsId)) {
+        // Throw StandByException,
+        // Clients could fail over and try another router.
+        if (rpcMonitor != null) {
+          rpcMonitor.proxyOpPermitRejected(nsId);
+        }
+        incrRejectedPermitForNs(nsId);
+        LOG.debug("Permit denied for ugi: {} for method: {}",
+            ugi, m.getName());
+        String msg =
+            "Router " + router.getRouterId() +
+                " is overloaded for NS: " + nsId;
+        throw new StandbyException(msg);
+      }
+      if (rpcMonitor != null) {
+        rpcMonitor.proxyOpPermitAccepted(nsId);
+      }
+      incrAcceptedPermitForNs(nsId);
+    }
+  }
```

Comment:
Here you can add a method acquirePermit(final String nsId, final UserGroupInformation ugi,
final String methodName, RouterRpcFairnessPolicyController controller) in RouterRpcClient, and then RouterRpcClient#acquirePermit(final String nsId, final UserGroupInformation ugi,
final RemoteMethod m, RouterRpcFairnessPolicyController controller) and RouterAsyncRpcClient#acquirePermit(final String nsId, final UserGroupInformation ugi,
final Method m, RouterRpcFairnessPolicyController controller) both call the public method.

Reply:
Nice suggestion, Have fixed.